File: python/strings/reverseWords.py
def reverseWords(s: str) -> str:
    ans = ""
    start = 0
    while start < len(s):
        space = start
        word = ""
        while space < len(s) and s[space] != " ":
            word += s[space]
            space += 1
        start = space + 1
        if word:
            ans = word + " " + ans
    return ans[:-1]


s = "  Cute you   are"
print(reverseWords(s))

File: cpp/strings/reverseWords.cpp
#include <bits/stdc++.h>
using namespace std;

string reverseWords(string s) {
  int i = 0;
  int n = s.size();
  string ans = "";
  ans.reserve(n);

  // find end of words and reverse it

  reverse(s.begin(), s.end());
  while (i < n) {
    // skip spaces
    while (i < n && s[i] == ' ')
      i++;

    int start = i;
    // find end of word
    while (i < n && s[i] != ' ')
      i++;

    reverse(s.begin() + start, s.begin() + i);
    ans += s.substr(start, i - start);
    ans += ' ';
  }
  while (ans[ans.size() - 1] == ' ')
    ans.erase(ans.size() - 1);

  return ans;
}

int main() {

  string s = "  a good   morning     ";

  cout << reverseWords(s);

  return 0;
}
